juju/juju
# Raft Changes for Including Vested Entities Against Pin Entries

## Description of change

This patch is the third in a series that will add the entity vested in a leadership pin/unpin operation to each pin. It depends upon https://github.com/juju/juju/pull/9330

It modifies the Raft FSM and store so that pin and unpin operations include the vested entity. In particular:
 - Pins are stored in memory as `map[lease.Key][]set.Tags`.
 - Snapshots correctly convert to/from this structure.
 - Lease expiry is suspended when there is no map entry, or its set is empty.

## QA steps

- Bootstrap to LXD.
- `juju deploy redis`.
- `juju add-unit redis`.
- `juju upgrade-series prepare 0` and confirm.
- `juju upgrade-series prepare 1` and confirm.
- The leader should be redis/0. Run `upgrade-series complete 0`
- Stop the container for machine 0 with `lxc stop {container ID}`.
- Watch `juju status` for a while to ensure that leadership does not change from redis/0, even though the machine is down.
- `juju upgrade-series complete 1`
- After a moment, leadership should change to redis/1.

Note that the pin/unpin feedback from the console remains unchanged. The pin/unpin will move from the client command in the forth and final patch in this series.

## Documentation changes

None.

## Bug reference

N/A
